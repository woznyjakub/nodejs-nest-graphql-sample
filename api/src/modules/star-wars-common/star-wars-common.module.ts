import { Module } from '@nestjs/common';

import { StarWarsCommonService } from './star-wars-common.service';

@Module({
  providers: [StarWarsCommonService],
  exports: [StarWarsCommonService],
})
export class StarWarsCommonModule {}
