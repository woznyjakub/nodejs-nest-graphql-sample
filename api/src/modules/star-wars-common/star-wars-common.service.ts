import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';

export type PagedResponse<Item> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Item[];
};

@Injectable()
export class StarWarsCommonService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  async fetchWithCaching<T extends object>(
    cacheKey: string,
    getterFn: () => Promise<T>,
  ): Promise<T> {
    const cachedData = await this.cacheManager.get<T>(cacheKey);
    if (cachedData) {
      return cachedData;
    }

    const data = await getterFn();
    await this.cacheManager.set(cacheKey, data);

    return data;
  }

  async getAllPages<T extends object>(
    cacheHandlerFn: (page: number) => string,
    getterFn: (page: number) => Promise<PagedResponse<T>>,
  ): Promise<T[]> {
    const limit = 25;
    let page = 1;

    const results: T[] = [];

    do {
      const data = await this.fetchWithCaching(cacheHandlerFn(page), () => getterFn(page));

      if (Array.isArray(data.results) || !results?.length) {
        results.push(...data.results);
      } else {
        break;
      }

      if (!data.next) {
        break;
      }
      page++;
    } while (page < limit);

    return results;
  }

  private buildCacheKey(
    resourceName: string,
    dataType: 'page' | 'id',
    value: string | number,
  ): string {
    return `swapi-${resourceName}-${dataType}-${value}`;
  }

  getVehicleCacheKey(id: number): string {
    return this.buildCacheKey('vehicle', 'id', id);
  }

  getVehiclesCacheKey(page: number): string {
    return this.buildCacheKey('vehicles', 'page', page);
  }

  getFilmCacheKey(id: number): string {
    return this.buildCacheKey('film', 'id', id);
  }

  getFilmsCacheKey(page: number): string {
    return this.buildCacheKey('films', 'page', page);
  }

  getStarshipCacheKey(id: number): string {
    return this.buildCacheKey('starship', 'id', id);
  }

  getStarshipsCacheKey(page: number): string {
    return this.buildCacheKey('starships', 'page', page);
  }

  getPlanetCacheKey(id: number): string {
    return this.buildCacheKey('planet', 'id', id);
  }

  getPlanetsCacheKey(page: number): string {
    return this.buildCacheKey('planets', 'page', page);
  }

  getSpeciesCacheKey(id: number): string {
    return this.buildCacheKey('species', 'id', id);
  }

  getSpeciesManyCacheKey(page: number): string {
    return this.buildCacheKey('speciesMany', 'page', page);
  }

  getPeopleCacheKey(page: number): string {
    return this.buildCacheKey('people', 'page', page);
  }
}
