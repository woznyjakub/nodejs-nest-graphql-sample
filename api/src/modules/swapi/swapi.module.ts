import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';

import { SwapiService } from './swapi.service';

@Module({
  imports: [HttpModule],
  providers: [SwapiService],
  exports: [SwapiService],
})
export class SwapiModule {}
