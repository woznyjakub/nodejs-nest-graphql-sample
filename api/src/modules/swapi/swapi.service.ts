import { HttpService } from '@nestjs/axios';
import { HttpException, Injectable } from '@nestjs/common';
import { ClassConstructor, plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { lastValueFrom } from 'rxjs';

import { FilmResponse, FilmsResponse } from './dto/film';
import { PeopleResponse } from './dto/people';
import { PlanetResponse, PlanetsResponse } from './dto/planet';
import { SpeciesManyResponse, SpeciesResponse } from './dto/species';
import { StarshipResponse, StarshipsResponse } from './dto/starship';
import { VehicleResponse, VehiclesResponse } from './dto/vehicle';

@Injectable()
export class SwapiService {
  constructor(private readonly httpService: HttpService) {}

  private readonly baseUrl = 'https://swapi.dev/api';

  private async getResource<T extends object>(
    name: string,
    dto: ClassConstructor<T>,
    id?: number,
    page?: number,
  ): Promise<T> {
    const url = `${this.baseUrl}/${name}${id ? `/${id}` : ''}`;
    const params = !id && page ? { page } : null;

    try {
      const response = await lastValueFrom(this.httpService.get(url, { params }));
      const resourceData: unknown = response.data;

      const resource = plainToInstance<T, unknown>(dto, resourceData, {
        enableImplicitConversion: true,
        excludeExtraneousValues: true,
      });

      const errors = await validate(resource);
      if (errors.length > 0) {
        throw new Error(
          `Swapi resource validation failed: ${errors.map((e) => e.toString()).join(', ')}`,
        );
      }

      return resource;
    } catch (error) {
      throw new HttpException(
        `Failed to get swapi resource (${name}). Error: ${error.message}`,
        error.status,
      );
    }
  }

  getVehicle(id: number): Promise<VehicleResponse> {
    return this.getResource('vehicles', VehicleResponse, id);
  }

  getVehicles(page: number = 1): Promise<VehiclesResponse> {
    return this.getResource('vehicles', VehiclesResponse, undefined, page);
  }

  getFilm(id: number): Promise<FilmResponse> {
    return this.getResource('films', FilmResponse, id);
  }

  getFilms(page: number = 1): Promise<FilmsResponse> {
    return this.getResource('films', FilmsResponse, undefined, page);
  }

  getStarship(id: number): Promise<StarshipResponse> {
    return this.getResource('starships', StarshipResponse, id);
  }

  getStarships(page: number = 1): Promise<StarshipsResponse> {
    return this.getResource('starships', StarshipsResponse, undefined, page);
  }

  getPlanet(id: number): Promise<PlanetResponse> {
    return this.getResource('planets', PlanetResponse, id);
  }

  getPlanets(page: number = 1): Promise<PlanetsResponse> {
    return this.getResource('planets', PlanetsResponse, undefined, page);
  }

  getSpecies(id: number): Promise<SpeciesResponse> {
    return this.getResource('species', SpeciesResponse, id);
  }

  getSpeciesMany(page: number = 1): Promise<SpeciesManyResponse> {
    return this.getResource('species', SpeciesManyResponse, undefined, page);
  }

  getPeople(page: number = 1): Promise<PeopleResponse> {
    return this.getResource('people', PeopleResponse, undefined, page);
  }
}
