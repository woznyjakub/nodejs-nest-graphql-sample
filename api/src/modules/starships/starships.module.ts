import { Module } from '@nestjs/common';

import { StarWarsCommonModule } from '../star-wars-common/star-wars-common.module';
import { SwapiModule } from '../swapi/swapi.module';

import { StarshipsResolver } from './starships.resolver';
import { StarshipsService } from './starships.service';

@Module({
  imports: [SwapiModule, StarWarsCommonModule],
  providers: [StarshipsResolver, StarshipsService],
})
export class StarshipsModule {}
