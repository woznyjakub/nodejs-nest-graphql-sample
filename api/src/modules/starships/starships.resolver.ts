import { Resolver, Query, Args, Int } from '@nestjs/graphql';

import { Starship } from './entities/starship.entity';
import { Starships } from './entities/starships.entity';
import { StarshipsService } from './starships.service';

@Resolver(() => Starship)
export class StarshipsResolver {
  constructor(private readonly vehiclesService: StarshipsService) {}

  @Query(() => Starships, { name: 'starships' })
  findAll(@Args('page', { type: () => Int }) page: number): Promise<Starships> {
    return this.vehiclesService.findAll(page);
  }

  @Query(() => Starship, { name: 'starship' })
  findOne(@Args('id', { type: () => Int }) id: number): Promise<Starship> {
    return this.vehiclesService.findOne(id);
  }
}
