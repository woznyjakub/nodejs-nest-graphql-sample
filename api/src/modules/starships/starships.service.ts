import { Injectable } from '@nestjs/common';

import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { StarshipResponse } from '../swapi/dto/starship';
import { SwapiService } from '../swapi/swapi.service';

import { Starship } from './entities/starship.entity';
import { Starships } from './entities/starships.entity';

@Injectable()
export class StarshipsService {
  constructor(
    private readonly swapiService: SwapiService,
    private readonly swCommonService: StarWarsCommonService,
  ) {}

  async findAll(page: number): Promise<Starships> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getStarshipsCacheKey(page),
      () => this.swapiService.getStarships(page),
    );

    return {
      count: swapiData.count,
      next: swapiData.next,
      previous: swapiData.previous,
      results: swapiData.results.map(this.mapSingleStarship),
    };
  }

  async findOne(id: number): Promise<Starship> {
    const swapiData = await this.swCommonService.fetchWithCaching(
      this.swCommonService.getStarshipCacheKey(id),
      () => this.swapiService.getStarship(id),
    );

    return this.mapSingleStarship(swapiData);
  }

  private mapSingleStarship(starship: StarshipResponse): Starship {
    return {
      name: starship.name,
      model: starship.model,
      manufacturer: starship.manufacturer,
      costInCredits: starship.cost_in_credits,
      length: starship.length,
      maxAtmospheringSpeed: starship.max_atmosphering_speed,
      crew: starship.crew,
      passengers: starship.passengers,
      cargoCapacity: starship.cargo_capacity,
      consumables: starship.consumables,
      hyperdriveRating: starship.hyperdrive_rating,
      MGLT: starship.MGLT,
      starshipClass: starship.starship_class,
      pilots: starship.pilots,
      films: starship.films,
      url: starship.url,
    };
  }
}
