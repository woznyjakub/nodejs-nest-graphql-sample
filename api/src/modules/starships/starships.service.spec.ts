import { Test, TestingModule } from '@nestjs/testing';

import { starshipResponseMock, starshipsResponseMock } from '../../test/mocks/swapi';
import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SwapiService } from '../swapi/swapi.service';

import { StarshipsService } from './starships.service';

describe('StarshipsService', () => {
  let starshipsService: StarshipsService;

  const mockFetchWithCaching = vi.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: StarWarsCommonService,
          useValue: {
            fetchWithCaching: mockFetchWithCaching,
          },
        },
        StarshipsService,
        {
          provide: SwapiService,
          useValue: {
            getStarship: vi.fn(),
            getStarships: vi.fn(),
          },
        },
      ],
    }).compile();

    starshipsService = module.get<StarshipsService>(StarshipsService);
  });

  describe('single starship data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(starshipResponseMock);

      const result = await starshipsService.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple starships data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(starshipsResponseMock);

      const result = await starshipsService.findAll(1);

      expect(result).toMatchSnapshot();
    });
  });
});
