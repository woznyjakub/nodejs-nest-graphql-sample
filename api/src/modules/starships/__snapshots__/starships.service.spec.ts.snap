// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StarshipsService > multiple starships data > should return mapped data 1`] = `
{
  "count": 36,
  "next": "https://swapi.dev/api/starships/?page=2",
  "previous": null,
  "results": [
    {
      "MGLT": "60",
      "cargoCapacity": "3000000",
      "consumables": "1 year",
      "costInCredits": "3500000",
      "crew": "30-165",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/",
      ],
      "hyperdriveRating": "2.0",
      "length": "150",
      "manufacturer": "Corellian Engineering Corporation",
      "maxAtmospheringSpeed": "950",
      "model": "CR90 corvette",
      "name": "CR90 corvette",
      "passengers": "600",
      "pilots": [],
      "starshipClass": "corvette",
      "url": "https://swapi.dev/api/starships/2/",
    },
    {
      "MGLT": "60",
      "cargoCapacity": "36000000",
      "consumables": "2 years",
      "costInCredits": "150000000",
      "crew": "47,060",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "2.0",
      "length": "1,600",
      "manufacturer": "Kuat Drive Yards",
      "maxAtmospheringSpeed": "975",
      "model": "Imperial I-class Star Destroyer",
      "name": "Star Destroyer",
      "passengers": "n/a",
      "pilots": [],
      "starshipClass": "Star Destroyer",
      "url": "https://swapi.dev/api/starships/3/",
    },
    {
      "MGLT": "70",
      "cargoCapacity": "180000",
      "consumables": "1 month",
      "costInCredits": "240000",
      "crew": "5",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "hyperdriveRating": "1.0",
      "length": "38",
      "manufacturer": "Sienar Fleet Systems, Cyngus Spaceworks",
      "maxAtmospheringSpeed": "1000",
      "model": "Sentinel-class landing craft",
      "name": "Sentinel-class landing craft",
      "passengers": "75",
      "pilots": [],
      "starshipClass": "landing craft",
      "url": "https://swapi.dev/api/starships/5/",
    },
    {
      "MGLT": "10",
      "cargoCapacity": "1000000000000",
      "consumables": "3 years",
      "costInCredits": "1000000000000",
      "crew": "342,953",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "hyperdriveRating": "4.0",
      "length": "120000",
      "manufacturer": "Imperial Department of Military Research, Sienar Fleet Systems",
      "maxAtmospheringSpeed": "n/a",
      "model": "DS-1 Orbital Battle Station",
      "name": "Death Star",
      "passengers": "843,342",
      "pilots": [],
      "starshipClass": "Deep Space Mobile Battlestation",
      "url": "https://swapi.dev/api/starships/9/",
    },
    {
      "MGLT": "75",
      "cargoCapacity": "100000",
      "consumables": "2 months",
      "costInCredits": "100000",
      "crew": "4",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "0.5",
      "length": "34.37",
      "manufacturer": "Corellian Engineering Corporation",
      "maxAtmospheringSpeed": "1050",
      "model": "YT-1300 light freighter",
      "name": "Millennium Falcon",
      "passengers": "6",
      "pilots": [
        "https://swapi.dev/api/people/13/",
        "https://swapi.dev/api/people/14/",
        "https://swapi.dev/api/people/25/",
        "https://swapi.dev/api/people/31/",
      ],
      "starshipClass": "Light freighter",
      "url": "https://swapi.dev/api/starships/10/",
    },
    {
      "MGLT": "80",
      "cargoCapacity": "110",
      "consumables": "1 week",
      "costInCredits": "134999",
      "crew": "2",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "1.0",
      "length": "14",
      "manufacturer": "Koensayr Manufacturing",
      "maxAtmospheringSpeed": "1000km",
      "model": "BTL Y-wing",
      "name": "Y-wing",
      "passengers": "0",
      "pilots": [],
      "starshipClass": "assault starfighter",
      "url": "https://swapi.dev/api/starships/11/",
    },
    {
      "MGLT": "100",
      "cargoCapacity": "110",
      "consumables": "1 week",
      "costInCredits": "149999",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "1.0",
      "length": "12.5",
      "manufacturer": "Incom Corporation",
      "maxAtmospheringSpeed": "1050",
      "model": "T-65 X-wing",
      "name": "X-wing",
      "passengers": "0",
      "pilots": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/18/",
        "https://swapi.dev/api/people/19/",
      ],
      "starshipClass": "Starfighter",
      "url": "https://swapi.dev/api/starships/12/",
    },
    {
      "MGLT": "105",
      "cargoCapacity": "150",
      "consumables": "5 days",
      "costInCredits": "unknown",
      "crew": "1",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "hyperdriveRating": "1.0",
      "length": "9.2",
      "manufacturer": "Sienar Fleet Systems",
      "maxAtmospheringSpeed": "1200",
      "model": "Twin Ion Engine Advanced x1",
      "name": "TIE Advanced x1",
      "passengers": "0",
      "pilots": [
        "https://swapi.dev/api/people/4/",
      ],
      "starshipClass": "Starfighter",
      "url": "https://swapi.dev/api/starships/13/",
    },
    {
      "MGLT": "40",
      "cargoCapacity": "250000000",
      "consumables": "6 years",
      "costInCredits": "1143350000",
      "crew": "279,144",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "2.0",
      "length": "19000",
      "manufacturer": "Kuat Drive Yards, Fondor Shipyards",
      "maxAtmospheringSpeed": "n/a",
      "model": "Executor-class star dreadnought",
      "name": "Executor",
      "passengers": "38000",
      "pilots": [],
      "starshipClass": "Star dreadnought",
      "url": "https://swapi.dev/api/starships/15/",
    },
    {
      "MGLT": "20",
      "cargoCapacity": "19000000",
      "consumables": "6 months",
      "costInCredits": "unknown",
      "crew": "6",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
      ],
      "hyperdriveRating": "4.0",
      "length": "90",
      "manufacturer": "Gallofree Yards, Inc.",
      "maxAtmospheringSpeed": "650",
      "model": "GR-75 medium transport",
      "name": "Rebel transport",
      "passengers": "90",
      "pilots": [],
      "starshipClass": "Medium transport",
      "url": "https://swapi.dev/api/starships/17/",
    },
  ],
}
`;

exports[`StarshipsService > single starship data > should return mapped data 1`] = `
{
  "MGLT": "60",
  "cargoCapacity": "3000000",
  "consumables": "1 year",
  "costInCredits": "3500000",
  "crew": "30-165",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/6/",
  ],
  "hyperdriveRating": "2.0",
  "length": "150",
  "manufacturer": "Corellian Engineering Corporation",
  "maxAtmospheringSpeed": "950",
  "model": "CR90 corvette",
  "name": "CR90 corvette",
  "passengers": "600",
  "pilots": [],
  "starshipClass": "corvette",
  "url": "https://swapi.dev/api/starships/2/",
}
`;
